### PR TITLE
Add distributed process group manager

### PR DESCRIPTION
**Pull request by Delaunay**

### CODE REVIEW

File: milatools/distributed/__init__.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    """Wrap the model to make it parallel if rank is not none"""
+    if rank() >= 0:
+        log.info("enabling multi-gpu %s", device_id())
+        return DistributedDataParallel(model, device_ids=[device_id()])
+
+    return model
+
+
+def record(fn, error_handler=None):
+    """Decorator that help record exception in a distrubted setup"""
```
Comment (breuleux) [inline milatools/distributed/__init__.py:177]:
```suggestion
    """Decorator that helps record an exception in a distributed setup"""
```

File: milatools/distributed/__init__.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    group = DistributedProcessGroup.INSTANCE
+    if group is None:
+        return default
+
+    return getattr(group, name)
+
+
+def barrier():
+    """block until all workers reach this"""
+    return __get_attr("barrier", lambda: lambda: True)()
```
Comment (breuleux) [inline milatools/distributed/__init__.py:117]: Did you mean the double lambda?

```suggestion
    return __get_attr("barrier", lambda: True)()
```

File: milatools/distributed/__init__.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    def device_id(self):
+        """Return the device id this script should use"""
+        if self.rank < 0:
+            return -1
+
+        return self.rank % torch.cuda.device_count()
+
+    @property
+    def device(self):
+        """Return the device this scrupt should use"""
```
Comment (breuleux) [inline milatools/distributed/__init__.py:100]:
```suggestion
        """Return the device this script should use"""
```

**Review by breuleux — COMMENTED** [review]
This could use some usage example to highlight the benefits.

Also, it should probably live under `milatools.torch.distributed` so that we have a fair organization of code for all frameworks (in case we want to add recipes for them).

**Comment by Delaunay** [comment]
> This could use some usage example to highlight the benefits.
> 
> Also, it should probably live under `milatools.torch.distributed` so that we have a fair organization of code for all frameworks (in case we want to add recipes for them).

Moved to torch folder

> This could use some usage example to highlight the benefits.

I added an example

**Comment by Delaunay** [comment]
We can make it even easier if we move some of the dataset downloading logic inside our dataset wrappers

**Pull request closed without merging**